### PR TITLE
[megatron] feat: add memory-efficient chunked top-k KL computation

### PR DESCRIPTION
**Pull request by HSYZhang**
### What does this PR do?

Adds an opt-in, memory-efficient chunked path for Megatron vocab-parallel top-k forward KL. It extends the FSDP chunking approach from #6593 to tensor-parallel vocabulary shards, avoiding full fp32 `[N, V_local]` log-softmax/probability buffers while preserving the existing outputs and gradients.

With chunked loss implemented, small student models (e.g., 4B or smaller) can reduce GPU memory via an FSDP-like chunked loss approach, eliminating the need to increase TP to lower peak memory usage. This prevents GPU idleness and significantly improves utilization. We previously used a 4B student model, but the trainer OOM'd during loss computation. The original approach required TP=4 to avoid OOM; with chunked loss enabled, we can now run at TP=1 without OOM, greatly improving GPU utilization and reducing idle time.

### Test

Results (b=2, s=96, V=2048, K=32; d_* denotes the maximum absolute error):

| Config | Comparison | d_loss | d_grad | d_smass | d_ocount | Verdict |
|---|---|---|---|---|---|---|
| fp32, clamp=-10 | baseline vs **reference** | 8.9e-8 | **1.2e-9** | — | — | ✅ |
| fp32, clamp=-10 | chunked(cs=7/64/4096) vs baseline | ≤1.2e-7 | ≤1.3e-9 | 1.9e-8 | 0 | ✅ |
| fp32, clamp=None | baseline vs reference | 8.9e-8 | 1.2e-9 | — | — | ✅ |
| fp32, clamp=None | chunked vs baseline | ≤1.2e-7 | ≤1.3e-9 | 1.9e-8 | 0 | ✅ |
| bf16, clamp=-10 | chunked vs baseline | ≤8.9e-8 | **1.9e-6** | 1.9e-8 | 0 | ✅ |

- Under fp32, chunked and baseline are nearly bit-for-bit identical (same floating-point operators and reduction order when TP=1);
- overlap_count is exactly identical (diff=0);
- For bf16 (storing bf16 logits and recomputing p in backward), the additional rounding is only ~1.9e-6;
- chunk_size=7 (many odd-sized chunks + a residual chunk) produces results consistent with 4096 → chunk boundary handling is correct.

Actual Qwen3.5-4B distillation dimensions: V=248320, K=128, TP=1, N=4096 tokens, clamp=-10.
Individually compared the maximum absolute/relative differences across 5 forward outputs + gradients.

**bf16 logits（actual online dtype）：**

| chunk_size | overlap_count diff | per_token_kl | student_mass | teacher_mass | overlap_adv | grad | Conclusion |
|---|---|---|---|---|---|---|---|
| 2048 | 0 | 0.00e+00 | 0.00e+00 | 0.00e+00 | 0.00e+00 | 0.00e+00 | **Bit-for-bit identical** |
| 4096 | 0 | 0.00e+00 | 0.00e+00 | 0.00e+00 | 0.00e+00 | 0.00e+00 | **Bit-for-bit identical** |
| 8192 | 0 | 0.00e+00 | 0.00e+00 | 0.00e+00 | 0.00e+00 | 0.00e+00 | **Bit-for-bit identical** |

**fp32 logits（stricter rounding observation）：**

| chunk_size | overlap_count diff | per_token_kl (abs/rel) | grad (abs/rel) | worst_rel | Conclusion |
|---|---|---|---|---|---|
| 2048 | 0 | 9.31e-10 / 8.13e-08 | 8.53e-14 / 2.96e-10 | 8.22e-07 | OK |
| 4096 | 0 | 9.31e-10 / 8.13e-08 | 8.53e-14 / 2.96e-10 | 8.22e-07 | OK |
| 8192 | 0 | 9.31e-10 / 8.13e-08 | 8.53e-14 / 2.96e-10 | 8.22e-07 | OK |

#### Peak memory

End-to-end forward/backward peak memory with `chunk_size=4,096`:

**TP=1 (`V_local=248,320`)**

| tokens (`N`) | logits size | baseline peak | chunked peak | saved | reduction |
|---:|---:|---:|---:|---:|---:|
| 4,096 | 1.90G | 13.28 GiB | 13.29 GiB | -0.01 GiB | -0.1% |
| 8,192 | 3.80G | 26.56 GiB | 18.97 GiB | 7.59 GiB | 28.6% |
| 16,384 | 7.60G | 53.12 GiB | 30.36 GiB | 22.76 GiB | 42.8% |
| 24,576 | 11.40G | 79.68 GiB | 45.54 GiB | 34.14 GiB | 42.8% |

**TP=4 (`V_local=62,080` per rank, matching the E20 launch)**

| tokens (`N`) | logits size per rank | baseline peak | chunked peak | saved per rank | reduction |
|---:|---:|---:|---:|---:|---:|
| 16,384 | 1.92G | 13.33 GiB | 7.63 GiB | 5.71 GiB | 42.8% |
| 32,768 | 3.84G | 26.67 GiB | 15.25 GiB | 11.41 GiB | 42.8% |
| 49,152 | 5.75G | 40.00 GiB | 22.88 GiB | 17.12 GiB | 42.8% |

At small token counts the chunk workspace makes peak memory effectively unchanged. The benefit grows with sequence length, reaching a stable `42.8%` reduction; for the TP=4 E20-scale case at `N=49,152`, this saves `17.12 GiB` per rank.

### API and Usage Example

No new API is introduced; this uses the existing distillation configuration added for the FSDP path:

```yaml
distillation_loss:
  use_chunked_topk: true
  chunked_topk_chunk_size: 4096
```

The feature remains disabled by default.

### Design & Code Changes

- Forward processes tokens in chunks, performs TP `MAX`/`SUM` reductions to compute the global `logZ`, and gathers only teacher top-k plus local student top-k statistics.
- Backward saves the original low-precision logits and per-token fp32 `logZ`, then recomputes probabilities chunk by chunk.
- This removes the persistent full fp32 probability tensor and limits temporary fp32 vocabulary storage to `[chunk_size, V_local]`, at the cost of additional TP collectives and recomputation.
- The baseline path is unchanged when chunking is disabled.

### Checklist Before Submitting

- [x] Searched for similar PRs; this implementation follows #6593.
- [x] PR title follows the required format.
- [x] Added targeted numerical validation, including chunk boundaries and realistic model dimensions.
- [x] No documentation update is required; the existing FSDP configuration is reused without API changes.

### CODE REVIEW

**Comment by tardis-key** [comment]
LGTM. Please check CI to confirm there are no failures caused by this PR. The pre-commit failures need you to update the code accordingly. @HSYZhang

**Comment by HSYZhang** [comment]
> LGTM. Please check CI to confirm there are no failures caused by this PR. The pre-commit failures need you to update the code accordingly. @HSYZhang

Thanks, done — I’ve fixed the pre-commit failures and rebased the latest code.
